src/Microsoft.AspNetCore.Mvc.ApiExplorer/ProblemAttributeApiDescriptionProvider.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    public class ProblemAttributeApiDescriptionProvider : IApiDescriptionProvider
+    {
+        private readonly IModelMetadataProvider _modelMetadaProvider;
+
+        public ProblemAttributeApiDescriptionProvider(IModelMetadataProvider modelMetadataProvider)
+        {
+            _modelMetadaProvider = modelMetadataProvider;
+        }
+
+        /// <remarks>
+        /// The order is set to execute after the <see cref="DefaultApiDescriptionProvider"/>.
+        /// </remarks>
+        public int Order => -1000 + 10;
+
+        public void OnProvidersExecuted(ApiDescriptionProviderContext context)
+        {
+        }
+
+        public void OnProvidersExecuting(ApiDescriptionProviderContext context)
+        {
+            foreach (var apiDescription in context.Results)
+            {
+                if (!apiDescription.ActionDescriptor.Properties.ContainsKey(ProblemDescriptionApplicationModelProvider.ProblemDescriptionAttributeKey))
+                {
+                    continue;
+                }
+
+                var parameters = apiDescription.ActionDescriptor.Parameters.Concat(apiDescription.ActionDescriptor.BoundProperties);
+                if (parameters.Any())
+                {
+                    apiDescription.SupportedResponseTypes.Add(CreateProblemResponse(StatusCodes.Status400BadRequest));
+
+                    if (parameters.Any(p => string.Equals("id", p.Name, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        apiDescription.SupportedResponseTypes.Add(CreateProblemResponse(StatusCodes.Status404NotFound));
```

Comment:
Shouldn't it also depend on the HTTP method? If I send a PUT request with an `id`, I'm not sure what 404 would mean in that case. Does our `ApiDescriptionProviderContext` have some way to say that certain `ApiResponseType` values only apply to certain HTTP methods?

Also, what would be the design for how to say your action does *not* ever return a 404?

Reply:
> Does our ApiDescriptionProviderContext have some way to say that certain ApiResponseType values only apply to certain HTTP methods?

Unfortunately not (and AFAIK neither does Swagger 2.0). 

> Also, what would be the design for how to say your action does not ever return a 404?

I intentionally left out any way to configure behavior at this point. We can do this in a follow up iteration. But I agree, ideally you want to be able to describe these in some way that affects both the Api description and the filter.